completions/bfs.bash

```diff
@@ -0,0 +1,170 @@
+
+  # arguments whose value is a dirname
+  local options_dircomp='-f'
+
+  # options with no value
+  local flags='-H -L -P -E -X -d -s -x -O1 -O2 -O3 -not -and -or -exclude -color -nocolor -daystart -depth -follow -ignore_readdir_race -noignore_readdir_race -mount -nohidden -status -unique -warn -nowarn -xdev -acl -capable -empty -executable -readable -writable -false -true -hidden -nogroup -nouser -sparse -xattr -delete -rm -exit -ls -print -print0 -printx -prune -quit -version -help'
+
+  local all_options="${flags} ${options_special} ${options_nocomp} ${options_filecomp} ${options_dircomp}"
+
+  # completions which require parsing the whole command string
+  case "${COMP_WORDS[*]}" in
+    (*\ -exec\ *|*\ -execdir\ *|*\ -ok\ *|*\ -okdir\ *)
+      case "$prev" in
+        (-exec|-execdir|-ok|-okdir)
+          # complete commands for first word after exec-type option
+          COMPREPLY=($(compgen -c -- "$cur"))
+          return
+        ;;
+      esac
+      # default completion only after exec-type option
+      #TODO: complete whatever command was provided after -exec
+      #TODO: detect end of -exec string and continue completing as normal
+      COMPREPLY=($(compgen -o default -o bashdefault -- "$cur"))
+      return
+    ;;
+  esac
+
+  # completions with 2-word lookbehind
+  case "$prev1" in
+    (-fprintf)
+      # -fprintf FORMAT FILE
+      #     Like -ls/-print/-print0/-printf, but write to FILE instead of standard
+      #     output
+      # when -fprintf is prev1, current word is FILE; perform file completion
+      COMPREPLY=($(compgen -f -- "$cur"))
+      return
+    ;;
+  esac
+
+  # completions with 1-word lookbehind
+  case "$prev" in
+    (-maxdepth|-mindepth|-amin|-Bmin|-cmin|-mmin|-asince|-Bsince|-csince|-msince|-atime|-Btime|-ctime|-mtime|-ilname|-iname|-ipath|-iregex|-iwholename|-inum|-links|-lname|-name|-path|-wholename|-regex|-since|-size|-used|-printf|-newerat|-newerBt|-newerct|-newermt|-xattrname)
+      # arguments whose values should not be completed
+      # (e.g. because they are numeric, glob, regexp, time, etc.)
+      COMPREPLY=()
+      return
+    ;;
+    (-anewer|-Bnewer|-cnewer|-mnewer|-newer|-samefile|-fls|-fprint|-fprint0|-neweraa|-neweraB|-newerac|-neweram|-newerBa|-newerBB|-newerBc|-newerBm|-newerca|-newercB|-newercc|-newercm|-newerma|-newermB|-newermc|-newermm)
+      # arguments whose value is a filename
+      COMPREPLY=($(compgen -f -- "$cur"))
+      return
+    ;;
+    (-f)
+      # arguments whose value is a dirname
+      COMPREPLY=($(compgen -d -- "$cur"))
+      return
+    ;;
+    (-type|-xtype)
+      # -type [bcdlpfswD]
+      #     Find files of the given type
+      # -xtype [bcdlpfswD]
+      #     Find files of the given type, following links when -type would not, and
+      #     vice versa
+      if [[ -n $cur ]] && ! [[ $cur =~ ,$ ]]; then
+        cur+=,
+      fi
+      COMPREPLY=("${cur}b" "${cur}c" "${cur}d" "${cur}l" "${cur}p" "${cur}f" "${cur}s" "${cur}w" "${cur}D")
```

Comment:
This can be shortened with brace expansion: `COMPREPLY=("${cur}"{b,c,d,l,p,f,s,w,D})`

Reply:
Didn't know you could do that, thanks!